docs/notebooks/distributed/custom_loop_pmap_example.md

```diff
@@ -0,0 +1,335 @@
+2. Connect to the runtime and run all cells. 
+
+NOTE: this example can be easily adapted to support TPU pod slices (e.g. `--accelerator_type v3-32`) as well as hosts with one or more GPUs attached.
+
+```{code-cell}
+:cellView: form
+:id: LD0juyi9JmIZ
+
+#@markdown The number of optimization steps to perform:
+MAXITER = 100  #@param{type:"integer"}
+#@markdown The number of samples in the (synthetic) dataset:
+NUM_SAMPLES = 50000  #@param{type:"integer"}
+#@markdown The number of features in the (synthetic) dataset:
+NUM_FEATURES = 784  #@param{type:"integer"}
+#@markdown The number of classes in the (synthetic) dataset:
+NUM_CLASSES = 10  #@param{type:"integer"}
+#@markdown The stepsize for the optimizer (set to 0.0 to use line search):
+STEPSIZE = 0.0  #@param{type:"number"}
+#@markdown The line search approach (either `'zoom'` or `backtracking`), ignored if `STEPSIZE > 0.0`:
+LINESEARCH = 'zoom'  #@param{type:"string"}
+```
+
++++ {"id": "9cSpF3jQMezP"}
+
+# Imports and TPU setup
+
+```{code-cell}
+:id: AkeC41JXNTv0
+
+%%capture
+%pip install jaxopt flax
+```
+
+```{code-cell}
+:id: 6UypR4hUNbve
+
+import functools
+import time
+from typing import Any, Callable, Tuple, Union
+
+from absl import app
+from absl import flags
+
+from flax import jax_utils
+from flax.training import common_utils
+
+import jax
+import jax.numpy as jnp
+import jax.tools.colab_tpu
+
+import jaxopt
+
+import matplotlib.pyplot as plt
+
+import numpy as np
+
+from sklearn import datasets
+```
+
+```{code-cell}
+---
+colab:
+  base_uri: https://localhost:8080/
+id: iMxI2ZU0JnJO
+outputId: 762fb71d-802b-4be1-8dc7-0edd62ab36ca
+---
+jax.tools.colab_tpu.setup_tpu()
+jax.local_devices()
+```
+
++++ {"id": "8ZReXwUIO4Tt"}
+
+# Type aliases
+
+```{code-cell}
+:id: 5A_PBZ7OO5rO
+
+Array = Union[np.ndarray, jax.Array]
+```
+
++++ {"id": "gbqilnNSNpw1"}
+
+# Auxiliary functions
+A MWE of how all-reduce mean/sum OPs can be introduced into JAXopt solver's `update` method by overriding `jax.value_and_grad`. Note that more complex wrappers are of course possible.
+
+```{code-cell}
+:id: awxxl498NwMI
+
+def pmean(fun: Callable[..., Any], axis_name: str = 'b') -> Callable[..., Any]:
+  """Applies `jax.lax.pmean` across `axis_name` for all of `fun`'s outputs."""
+  maybe_pmean = lambda t: jax.lax.pmean(t, axis_name) if t is not None else t
+  @functools.wraps(fun)
+  def wrapper(*args, **kwargs):
+    return jax.tree_map(maybe_pmean, fun(*args, **kwargs))
+  return wrapper
+```
+
++++ {"id": "clQeXb77ObhK"}
+
+A small utility to shard `Array`s across the available devices:
+
+```{code-cell}
+:id: rMntHIEFOeyO
+
+def shard_array(array: Array) -> jax.Array:
+  """Shards `array` along its leading dimension."""
+  return jax.device_put_sharded(
+      shards=list(common_utils.shard(array)),
+      devices=jax.devices())
+```
+
++++ {"id": "4Pph9LUAOmhi"}
+
+# Custom-loop
+Shows a comparison of how single-device and data-parallel SPMD differ in a vanilla JAXopt custom training loop
```

Comment:
could we replace SPMD by a term that's less technical/intimidating? 

Also, it would be nice to be a bit more explicit here on where is the comparison since the actual comparison won't happen until (relatively) much later. How about something like:

The following code uses data-parallelism in the train loop. Through the `use_pmap` keyword argument we can deactivate this parallelism. We'll use this feature later to benchmark the impact of parallelism.

Reply:
Thanks Fabian, your phrasing is so much better. Done!